### ISSUE TITLE
Don't assume connection error is the only error that happened

### ISSUE THREAD

**Issue by jzabroski** [chucknorris/roundhouse#209]
https://github.com/chucknorris/roundhouse/commit/3bbefb4a48c62a89044a523b01089ba66ce8a934#commitcomment-12587865

**Comment by jzabroski** [comment]
Just to add a suggestion, I would make this re-try logic completely optional, via a command line argument to rh.exe

@ferventcoder - thoughts?

**Comment by ChrisMissal** [comment]
Wow @jzabroski, that's a nasty one! Sorry you had to deal with it, and thanks for reporting it.

The intended behavior is meant to be:

> If an error occurs due to a connection exception (error number 233), re-try one time.

I think we can avoid the opt-in re-try logic if the code is changed to check that all the errors in the collection are the connection exception error number of 233. If any error in the exception was not a connection problem, the exception would be thrown.

To me this is more correct, and I believe it would solve the issues that were plaguing you.

**Comment by ferventcoder** [comment]
Agreed, we should look only for the connection exception. And yikes on the above!

**Comment by jzabroski** [comment]
@ferventcoder  I think you meant to say "should NOT only look".  The problem is that the previous commands would roll back, and then you are continuing with the rest of RH's script sequence.

**Comment by ferventcoder** [comment]
I think what you mean is what I mean. Haha

**Comment by jzabroski** [comment]
Here is another way this bug can crop up.

1. Run RH as a user other than db_owner.
2. Run an Up script 0001.sql without fully qualified database names, e.g. `select * from Person` instead of `select * from dbo.Person`
3. Run an Up script 0002.sql with fully qualified database names

2 will fail, cause a retry, and proceed to step 3.

**Comment by jzabroski** [comment]
@ChrisMissal @ferventcoder 

After doing some research, it appears the only reliable and general way to detect "network errors" and differentiate them from all other SqlException objects is to check for the [SqlConnection.ClientConnectionId](https://msdn.microsoft.com/en-us/library/system.data.sqlclient.sqlconnection.clientconnectionid(v=vs.110).aspx) in the [Connectivity Ring Buffer](http://blogs.msdn.com/b/sql_protocols/archive/2008/05/20/connectivity-troubleshooting-in-sql-server-2008-with-the-connectivity-ring-buffer.aspx).

FYI: I _also_ looked into the Lokad project, mentioned in the comments of the offending code in AdoNetDatabase.cs.  It seems like they are trying to solve particular connectivity issues with sending long running commands to SQL Server instances hosted in Azure.  Please confirm.

**Comment by jzabroski** [comment]
After thinking about this a bit more deeply, there is a _slight_ chance the connection is lost because the whole machine went down.  In which case, the whole Connectivity Ring Bugger dynamic manangement view data would be lost.

Since @@SPID is just a smallint, you can't just filter by the SPID of the SqlConnection - you need to use as many fields in the XML Record as possible to preclude possibility of the machine restarting, such as verifying the connection was made by the machine RH is running on.  Because on a very busy cloud-based SQL Server instance, SPID will get re-used very quickly.

Thoughts?

**Comment by jzabroski** [comment]
I tried to write some code that would query the OS ring buffer dynamic management view, but it just didn't seem to work right.  I am probably going to have to break this ticket up into two things - one, respect the --trx flag, and two, do proper retries.
